sstables/sstable_set.hh

```diff
@@ -24,84 +24,207 @@
 #include "shared_sstable.hh"
 #include "dht/i_partitioner.hh"
 #include <seastar/core/shared_ptr.hh>
+#include <seastar/core/weak_ptr.hh>
 #include <vector>
 
 namespace sstables {
 
-class sstable_set_impl;
-class incremental_selector_impl;
+struct versioned_sstable_data {
+    shared_sstable sst;
+    uint64_t version_added;
+    std::optional<uint64_t> version_erased;
+};
+
+using versioned_sstable = seastar::lw_shared_ptr<versioned_sstable_data>;
+
+class incremental_selector_impl {
+public:
+    virtual ~incremental_selector_impl() {}
+    virtual std::tuple<dht::partition_range, std::vector<versioned_sstable>, dht::ring_position_view> select(const dht::ring_position_view&) = 0;
+};
+
+class sstable_set_impl {
+public:
+    virtual ~sstable_set_impl() {}
+    virtual std::vector<versioned_sstable> select(const dht::partition_range& range) const = 0;
+    virtual void insert(versioned_sstable sst) = 0;
+    virtual void erase(versioned_sstable sst) = 0;
+    virtual std::unique_ptr<incremental_selector_impl> make_incremental_selector() const = 0;
+};
 
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable>&&);
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class sstable_set_snapshot;
+class version_reference;
+
+class sstable_list {
+    lw_shared_ptr<const std::unordered_set<versioned_sstable>> _versioned_sstables;
+    lw_shared_ptr<std::unordered_set<shared_sstable>> _unversioned_sstables;
```

Comment:
_versioned_sstables is the set of all sstables from all versions that are stored in the versioned sstable_set. 
In all but one cases the sstable_list is used as a reference to the set of all sstables in an sstable_set_snapshot - we have _version to be able to iterate only over the sstables that are alive in the corresponding version. The sstable_list returned by table::get_sstables_including_compacted_undeleted() may not be such a reference though - it may be a reference to a new set of sstables that includes all sstables from a snapshot and some others, that weren't added or erased in any version - _unversioned_sstables. 
Before the changes, the exceptional case was handled by copying the entire set of sstables from the sstable_set and adding new sstables to the copy. We could do a similar thing here, but considering we would still want to handle the case where sstable_list is just a reference to the set of sstables from a snapshot, we would need to keep the _versioned_sstables field, but in the exceptional case it could point to the copy of the original set, and we could add new sstables with artificial versions to the copy.
I think the other solution is equally confusing though, and has a worse complexity.

Reply:
So, `_unversioned_sstables` are those sstables that are in *all* currently live versions, and `_versioned_sstables` are those that are added/removed in a certain version. Makes sense, however the naming is very misleading, it makes one think that we have different version of the same sstable.

Also, why are all these containers `lw_shared_ptr`? This observation applies to the entire patch -- in general I have a feeling that we have way more shared pointers than necessary. I'd like to emphasize that this is just a gut feeling, in the current form I have a hard time forming a big picture of all the changes and therefore can't rely on understanding of the code.

Reply:
I'll change the naming in the next patch, but I think I should clarify what these members mean before that. You are right about the meaning of _versioned_sstables, but it seems I didn't explain the _unversioned_sstables clearly. The unversioned sstables are sstables that were added to the sstable_list but don't appear in any of the versions of the sstable_set. It looks like you're right about not needing to keep a lw_shared_ptr to them. 
We're keeping lw_shared_ptrs to sstable_lists because that's basically what we were doing before the change and I didn't want to make too many unnecessary changes